### DISCUSSION TITLE
Three issues with the MGCB editor GUI

### DISCUSSION THREAD

**Discussion by AugsEU** [MonoGame/MonoGame#8152] [Content]
Hi,

I have been working with monogame for a while and there are a few things about the mgcb editor that annoy me to no end:

* The default view for some reason prioritises the build log and makes the tree view tiny. The tree view is the main part of this GUI so it should be the biggest
* PNGs default to have "Premultiply" enabled. This is wrong as **no** PNG should be premultiplied. "if a PNG image is premultiplied, it is not a compilant PNG" -Lee Daniel Crocker. It should default to disabled!!
* If a file doesn't have an imported/processor it will still default to "Build" with "Invalid / Missing Importer". This should just default to copy. That way I don't have to manually set "Copy" every time I want to import a .txt or .xml file.

I would like to get a build of the source code for this GUI so I can make these changes myself. I am not sure where the source code is or how I can get a local version building? Could someone help me with that please?

**Comment by KelsamGames** [comment]
> I am not sure where the source code is or how I can get a local version building? Could someone help me with that please?

https://github.com/MonoGame/MonoGame/tree/develop/Tools/MonoGame.Content.Builder.Editor

**Comment by AristurtleDev** [comment]
>PNGs default to have "Premultiply" enabled. This is wrong as no PNG should be premultiplied. "if a PNG image is premultiplied, it is not a compilant PNG" -Lee Daniel Crocker. It should default to disabled!!

As to this point, I would hazard a guess that's it's set to premuliply enabled by default because the `SpriteBatch.Begin()` method uses `BlendState.AlphaBlend` if one isn't provided in the parameters

https://github.com/MonoGame/MonoGame/blob/a6ef14b6db25c9fced6157bbc2270255aa07448c/MonoGame.Framework/Graphics/SpriteBatch.cs#L93-L100

**Reply by ThomasFOG** [reply]
Yes, historically XNA has been pushing for premultiplied alpha because it is a more optimized form of alpha blending calculation and the stock shaders make use of it.

Also, this has nothing to do with the input format (PNG). There's a misunderstanding regarding what the content pipeline is. It is not designed to keep the original file's properties, but to create an output format agnostically of the input format.

**Reply by BugInfestedCode** [reply]
>.....**_create an output format agnostically of the input format_**.

Could you elaborate on what is being done by the content pipeline to the inputted files?

Thank you.

**Reply by AristurtleDev** [reply]
> Could you elaborate on what is being done by the content pipeline to the inputted files?

The content pipeline is performing the following

## Texture Importing
The TextureImporter reads the color data from the input format and creates a `BitmapContent` object which contains the data about the image like the Width, Height, and Pixel data.  This part essentially converts the input image format into an agnostic format that is then passed to the `TextureProcessor`


## Texture Processing
The `TextureProcessor` takes the data given and optimizes the pixel data for the platform being targeted.  It calls the [`TextureProfile.CreateForPlatform`](https://github.com/MonoGame/MonoGame/blob/06c49f2d32f2a6a121ffbaf8f63ea914e7bb61fa/MonoGame.Framework.Content.Pipeline/Graphics/TextureProfile.cs#L21) method which will create a `TextureProfile` instance that is used to translate the agnostic format given from the importer into a platform specific format depending on the platform being targeted, such as Pvr for iOS, Etc for Android, and DXT for desktops.

https://github.com/MonoGame/MonoGame/blob/06c49f2d32f2a6a121ffbaf8f63ea914e7bb61fa/MonoGame.Framework.Content.Pipeline/Graphics/DefaultTextureProfile.cs#L45-L79

The result of this processing to a platform specific format is what is output into the .xnb file that is produced.  So when you then load it in your game using the `ContentManager.Load<Texture2D>` method, it's not loading the .png or .jpg that you added, it's instead loading the preprocessed optimized format

**Reply by BugInfestedCode** [reply]
@AristurtleDev, Thank you for the informative answer.
